Improve performance on json to psobject conversion

powercode opened the pull request:
## PR Summary

Changes to the `JsonObject` class to make conversions from json to `PSObject`s faster.

Three main changes:
Convert from JArray directly to object instead of creating a list and do list.ToArray().
Do not use PSObject to check for existing members, since that is a very slow and allocation heavy (sic!)  code path.
PreAllocation the members in PSObject by using the newly added constructor accepting an initial member count.

This amounts to a speed-up in the order of 7x.

Cmdlets that benefit are `Convertfrom-Json` and `Invoke-RestMethod`. 


## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

powercode committed:
Code cleanup and simplification

Using pattern matching to simplify code.
Removing some redundant initializations.
nameof(input) instead of "input" in exception.
Simplification of assignment
Removing code redundancies

powercode committed:
Using array directly instead of list + .ToArray() to reduce allocations.

powercode committed:
Use a hashset as a faster method of tracking member duplicates.

powercode committed:
CodeFactor fixes

markekraus commented:
```powershell
$json = @'
{
    "id": "string",
    "data": [
        {                       
        "accountId": 0,
        "productId": 0,
        "resourceLocationId": 0,
        "consumedServiceId": 0,
        "departmentId": 0,
        "accountOwnerEmail": "string",
        "accountName": "string",
        "serviceAdministratorId": "string",
        "subscriptionId": 0,
        "subscriptionGuid": "string",
        "subscriptionName": "string",
        "date": "2017-04-27T23:01:43.799Z",
        "product": "string",
        "meterId": "string",
        "meterCategory": "string",
        "meterSubCategory": "string",
        "meterRegion": "string",
        "meterName": "string",
        "consumedQuantity": 0,
        "resourceRate": 0,
        "Cost": 0,
        "resourceLocation": "string",
        "consumedService": "string",
        "instanceId": "string",
        "serviceInfo1": "string",
        "serviceInfo2": "string",
        "additionalInfo": "string",
        "tags": "string",
        "storeServiceIdentifier": "string",
        "departmentName": "string",
        "costCenter": "string",
        "unitOfMeasure": "string",
        "resourceGroup": "string"
        }
    ],
    "nextLink": "string"
}
'@
$json = 1..1000 | % { $json }
$json = '[' + ($json -join ',') + ']'
Measure-Command {$Json | ConvertFrom-Json } | % TotalMilliseconds
````

Runs in 497.8867 miliseconds on the most recent build in this PR vs 1104.9008 ms in preview.4 and 318.1642ms in 5.1. Much better!! Thanks @powercode

markekraus reviewed (approved):
LGTM

powercode commented:
Did you test that with a crossgen build?

I get a lot better numbers.
Same code but with 20000 iterations - I get 27091 ms on 6.0.3 and 3635 ms on the PR.

Curious about the difference.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -142,10 +134,10 @@ private static PSObject PopulateFromJDictionary(JObject entries, out ErrorRecord
                 {
-                    string errorMsg = string.Format(CultureInfo.InvariantCulture,
-                                                    WebCmdletStrings.DuplicateKeysInJsonString, entry.Key);
+                    var errorMsg = string.Format(CultureInfo.InvariantCulture, WebCmdletStrings.DuplicateKeysInJsonString, entry.Key);
```
minor: `CurrentCulture` should be used when constructing error message. A flaw in the original code.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -83,55 +86,44 @@ public static object ConvertFromJson(string input, bool returnHashTable, out Err
-                    }
+                    obj = returnHashTable ?
+                              PopulateHashTableFromJDictionary(dictionary, out error) :
+                              PopulateFromJDictionary(dictionary, memberTracker, out error);
```
When reaching here, we already know the size of the `JObject`, so maybe we can replace `memberTracker` with `new DuplicateMemberSet(dictionary.Count)`? (just like what you did in other places)
Then `DefaultMemberSetCapacity` won't be needed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -83,55 +86,44 @@ public static object ConvertFromJson(string input, bool returnHashTable, out Err
                 {
-                    string errorMsg = string.Format(CultureInfo.InvariantCulture,
-                                                    WebCmdletStrings.EmptyKeyInJsonString);
+                    var errorMsg = string.Format(CultureInfo.InvariantCulture, WebCmdletStrings.EmptyKeyInJsonString);
```
minor: `CurrentCulture` should be used when constructing error strings. It's a flaw in the original code.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -156,11 +148,9 @@ private static PSObject PopulateFromJDictionary(JObject entries, out ErrorRecord
                 {
-                    string errorMsg = string.Format(CultureInfo.InvariantCulture,
-                                                    WebCmdletStrings.KeysWithDifferentCasingInJsonString, property.Name, entry.Key);
+                    var errorMsg = string.Format(CultureInfo.InvariantCulture, WebCmdletStrings.KeysWithDifferentCasingInJsonString, propertyName, entry.Key);
```
minor: `CurrentCulture` should be used when constructing error message.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -255,8 +246,7 @@ private static Hashtable PopulateHashTableFromJDictionary(JObject entries, out E
                 {
-                    string errorMsg = string.Format(CultureInfo.InvariantCulture,
-                                                    WebCmdletStrings.DuplicateKeysInJsonString, entry.Key);
+                    string errorMsg = string.Format(CultureInfo.InvariantCulture, WebCmdletStrings.DuplicateKeysInJsonString, entry.Key);
```
minor: same here

daxian-dbw commented:
Good perf work @powercode! The changes look good to me in general. Left a few minor comments.

powercode committed:
Addressing dongbo's comments.

daxian-dbw reviewed (approved):
Ship it! 🎉

powercode commented:
I looked a bit at the difference in perf to Windows PowerShell. 

`Newtonsoft.Json.JsonConvert.DeserializeObject` is slower than `System.Web.Script.Serialization.JavaScriptObjectDeserializer`. Looking at the code to see if I can figure out what the root cause is.

dantraMSFT reviewed (commented):
Other than the minor suggestion, LGTM.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -83,55 +83,44 @@ public static object ConvertFromJson(string input, bool returnHashTable, out Err
+                              PopulateHashTableFromJDictionary(dictionary, out error) :
+                              PopulateFromJDictionary(dictionary, new DuplicateMemberSet(dictionary.Count), out error);
                 }
                 else
```
I'm moving the return into the try statement so I get 

```csharp

if (object isJObject dictionary)
{...}
else if (obj is JArray list)
{///}
// if obj is a anything but JArray or JObject, we just return as is
return obj;
```

powercode replied:
Then it's clear anyway that we return obj as is for the other cases.


Or how about 

```csharp
switch (obj)
{
    case JObject dictionary:
        return returnHashTable
                   ? PopulateHashTableFromJDictionary(dictionary, out error)
                   : PopulateFromJDictionary(dictionary, new DuplicateMemberSet(dictionary.Count), out error);
    case JArray list:
        return returnHashTable
                   ? PopulateHashTableFromJArray(list, out error)
                   : PopulateFromJArray(list, out error);
    default: return obj;
}
```
?

powercode replied:
That makes the alternative really clear.

powercode committed:
addressing dantraMSFT comments - part 2

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -142,10 +124,10 @@ private static PSObject PopulateFromJDictionary(JObject entries, out ErrorRecord
                 // does not throw when encountering duplicates and just uses the last entry.
-                if (result.Properties.Any(psPropertyInfo => psPropertyInfo.Name.Equals(entry.Key, StringComparison.InvariantCulture)))
+                if (memberTracker.TryGetValue(entry.Key, out var maybePropertyName)
+                    && string.Compare(entry.Key, maybePropertyName, StringComparison.CurrentCulture) == 0)
```
We throw different errors in different cases. A duplicate key with same casing is a hard "NO", while a duplicate key with different casing is a workaround-able.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -170,78 +150,79 @@ private static PSObject PopulateFromJDictionary(JObject entries, out ErrorRecord
-                    JValue theValue = entry.Value as JValue;
+                    // Value
+                    JValue theValue = (JValue)entry.Value;
                     result.Properties.Add(new PSNoteProperty(entry.Key, theValue.Value));
```
I'm fine with using casting when we are for sure expecting the cast-to type.
If the value does happen to be a different type, then it's a bug that needs to be fixed, and casting operator will make the bug to surface easily (while `as` in this case might make the bug harder to be noticed).

iSazonov replied:
> it's a bug that needs to be fixed

@daxian-dbw If I remember right in previous discussion we concluded that we don't want an exception here. So I ping @markekraus to confirm or reject this.

Also we could consider the change as out of the PR and revert and open a tracking issue.

markekraus replied:
@iSazonov I don't recall any reason why we need `as` over casting here.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -265,79 +245,79 @@ private static Hashtable PopulateHashTableFromJDictionary(JObject entries, out E
+                else
                 {
+                    // Value
                     JValue theValue = entry.Value as JValue;
```
Yeah, this is a bit inconsistent in this file.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -82,56 +81,39 @@ public static object ConvertFromJson(string input, bool returnHashTable, out Err
             error = null;
-            PSObject result = new PSObject();
+            var result = new PSObject(entries.Count);
             foreach (var entry in entries)
```
Why?

powercode replied:
The type is clearly stated by the ctor. Why repeat it?

Our coding guidelines has 

```csharp
var example = obj as Example;
```

as an example. I don't see how this is any different. There can be no confusion as to what type result has. 

In general, I think it would be better to consider types as something the compiler uses to ensure we don't use something in an unintended way. 

Personally, I find all the type annotations reducing readability, since they get in the way of the essence.

powercode replied:
Again, as background, I really recommend Herb Sutters https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

powercode replied:
Also, when looking at high quality, modern, C# code bases this is a very common, and in my eyes non-controversial.

https://github.com/dotnet/roslyn/blob/acd31862be412fcb4f7e0bc7a7db63ea8228a327/src/Compilers/Shared/DesktopBuildClient.cs#L39

https://github.com/dotnet/roslyn/blob/master/src/Compilers/Shared/BuildServerConnection.cs

iSazonov replied:
Do we both say about `foreach (var entry in entries) ` ?

powercode replied:
I don't think that the type of `entry` is of interest here. It is some entity that has a `Key` and a `Value`. Should that type change, nothing in the code would need to change, as long as it has a `Key` and a `Value`. 
That is captured nicely with the use of  `var`. 

I think we should only specify the type explicitly when it is important for some reason that the value is of a certain type. That way, it calls attention to it - something interesting is happening here since we explicitly specified the type. It may be that a variable is declared as a base type instead of the actual type returned by a function.

The C# coding conventions on [msdn](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions) has actually captured this:
> Use implicit typing for local variables when the type of the variable is obvious from the right side of the assignment, *or when the precise type is not important.*

> Use implicit typing to determine the type of the loop variable in for and foreach loops. 
The semantics would also change.

powercode commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -24,45 +19,49 @@ namespace Microsoft.PowerShell.Commands
-        private const int maxDepthAllowed = 100;
+        private class DuplicateMemberSet : HashSet<string>
+        {
+            public DuplicateMemberSet(int capacity) : base(capacity, StringComparer.CurrentCultureIgnoreCase)
```
I think you are correct that CurrentCultureIgnoreCase is wroing.  The previous code used the PSObject.Properties, and that uses OrdinalIgnoreCase, for it's member dictionary. 

Not sure if we should use InvariantCultureIgnoreCase or OrdinalIgnoreCase.

daxian-dbw replied:
I vote for `OrdinalIgnoreCase`.

iSazonov replied:
I agree about OrdinalIgnoreCase

iSazonov replied:
@powercode I think if we reference Hashtable class in comments we should use quotes like 'Hashtable' to bypass StypeCop warnings. As for variable names we should follow Camel style so returnHashTable is ok.

powercode committed:
Addressing review comments

Using OrdinalIgnoreCase for DuplicateMemberHashSet.

Renaming DuplicateMemberSet -> DuplicateMemberHashSet
Renaming returnHashTable -> returnHashtable

Changing help to not use the HashTable casing.

iSazonov commented:
Good catch! I agree with using a type switch.

powercode committed:
Type switch on JArray, JObject and JValue to more clearly specify intent.

Also specifying initial capacity on hashtable.

powercode committed:
Spacing around switch.

powercode commented:
@iSazonov  I'm getting code factor errors on 

```csharp

switch (x)
{
    case 1:
    {
        break;
    }
    case 2:
    {
        break;
    }
}
```

It seems to want it like 

switch (x)
{
    case 1:
    {
        break;
    }

    case 2:
    {
        break;
    }
}
```


That feels strange. Do you know if it is intentional or a side effect?

iSazonov reviewed (approved):
LGTM.

markekraus commented:
@powercode I believe that spacing rule is intentional. It is easier to read code when there is some white space separating the end of one code block and the beginning of another.

iSazonov commented:
@markekraus Thanks for review!
